File: backtracking/sudoku/sudoku_solver.go
package sudoku

/**
数字 1-9 在每一行只能出现一次。
数字 1-9 在每一列只能出现一次。
数字 1-9 在每一个以粗实线分隔的 3x3 宫内只能出现一次。
*/
// 这种解法是错误的，跟N皇后不一样，因为N皇后问题只需要在每一行中添加一个皇后，所以我们只需要一层for循环遍历一行，递归来遍历列
// 而解数独在一行的所有地方都要放上元素，所以是二维递归，他的递归树比N皇后要大得多
func solveSudoku(board [][]byte) {
	nums := []byte{'1', '2', '3', '4', '5', '6', '7', '8', '9'}
	// 错误的写法
	var backtrack func(row int)
	backtrack = func(row int) {
		if row == len(board) {
			return
		}

		// 列
		for i := 0; i < 9; i++ {
			if board[row][i] == '.' {
				for _, v := range nums {
					if isValid(board, v, row, i) {
						board[row][i] = v
						backtrack(row + 1)
						board[row][i] = '.'
					} else {
						continue
					}
				}
			}
		}
	}
	backtrack(0)
}

func isValid(board [][]byte, val byte, row, col int) bool {
	// 行里是否有重复
	for i := 0; i < 9; i++ {
		if board[row][i] == val {
			return false
		}
	}

	// 列里是否有重复
	for i := 0; i < 9; i++ {
		if board[i][col] == val {
			return false
		}
	}

	// 所属3×3的范围内是否有重复值
	row = (row / 3) * 3
	col = (col / 3) * 3
	for i := row; i <= row+2; i++ {
		for j := col; j <= col+2; j++ {
			if board[i][j] == val {
				return false
			}
		}
	}
	return true
}

func solveSudoku2(board [][]byte) {
	var backtracking func(board [][]byte) bool
	// 本题递归不用终止条件，解数独是要遍历整个树形结构寻找可能的叶子节点就立刻返回。
	backtracking = func(board [][]byte) bool {
		// 行
		for i := 0; i < 9; i++ {
			for j := 0; j < 9; j++ {
				if board[i][j] != '.' {
					continue
				}

				for v := '1'; v <= '9'; v++ {
					if isValid(board, byte(v), i, j) {
						board[i][j] = byte(v)
						if backtracking(board) {
							return true
						}
						board[i][j] = '.'
					}
				}
				// 从1-9每个数都试了一遍之后没有正确答案，返回false
				return false
			}
		}
		return true // 遍历完没有返回false，说明找到了合适棋盘位置了
	}
	backtracking(board)
}

File: backtracking/other/sudoku/SudokuSolver.java
package backtracking.other.sudoku;

/**
 * leetcode 37 hard 解数独
 * 
 * 编写一个程序，通过填充空格来解决数独问题。
 * 数独的解法需 遵循如下规则：
 *  数字 1-9 在每一行只能出现一次。
 *  数字 1-9 在每一列只能出现一次。
 *  数字 1-9 在每一个以粗实线分隔的 3x3 宫内只能出现一次。（请参考示例图）
 * 
 * 数独部分空格内已填入了数字，空白格用 '.' 表示。
 */
public class SudokuSolver {
    public void solveSudoku(char[][] board) {
        backtrack(board);
    }

    // private void backtrack(char[][] board, int row) {
    //     if (row == board.length) {
    //         return;
    //     }

    //     for (int col = 0; col < 9; col++) {
    //         if (board[row][col] == '.') {
    //             for (int i = 1; i <= 9; i++) {
    //                 if (isValid(board, row, col, i)) {
    //                     board[row][col] = (char) i;
    //                     backtrack(board, row + 1);
    //                     board[row][col] = '.';
    //                 }
    //             }
    //         }
    //     }
    // }

    // 在棋盘的i，j位置放上一个数num是否合理
    private boolean isValid(char[][] board, int i, int j, char num) {
        // 同一行的位置上
        for (int col = 0; col < 9; col++) {
            if (board[i][col] == num) {
                return false;
            }
        }

        // 同一列的位置上
        for (int row = 0; row < 9; row++) {
            if (board[row][j] == num) {
                return false;
            }
        }

        // 在3*3的区域内
        int X = (i / 3) * 3;
        int Y = (j / 3) * 3;

        for (int x = X; x <= X + 2; x++) {
            for (int y = Y; y <= Y + 2; y++) {
                if (board[x][y] == num) {
                    return false;
                }
            }
        }

        return true;
    }

    // 之所以设置成boolean，是相当于剪枝
    public boolean backtrack(char[][] board) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    for (char val = '1'; val <= '9'; val++) {
                        if (isValid(board, i, j, val)) {
                            board[i][j] = (char) val;
                            if (backtrack(board)) {
                                return true;
                            }
                            board[i][j] = '.';
                        }
                    }
                }
            }
        }

        return false;
    }
}
